executor/metric_reader.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/infoschema/metricschema"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+)
+
+const promReadTimeout = time.Second * 10
+
+// MetricReaderExec uses to read metric data.
+type MetricReaderExec struct {
+	baseExecutor
+	table      *model.TableInfo
+	tblDef     *metricschema.MetricTableDef
+	outputCols []*model.ColumnInfo
+	done       bool
+}
+
+// Next implements the Executor Next interface.
+func (e *MetricReaderExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if e.done {
+		return nil
+	}
+	rows, err := e.getRows()
+	if err != nil {
+		return err
+	}
+	e.done = true
+
+	if len(rows) == 0 {
+		return nil
+	}
+
+	req.GrowAndReset(len(rows))
+	mutableRow := chunk.MutRowFromTypes(retTypes(e))
+	for _, row := range rows {
+		mutableRow.SetDatums(row...)
+		req.AppendRow(mutableRow.ToRow())
+	}
+	return nil
+}
+
+func (e *MetricReaderExec) getRows() (fullRows [][]types.Datum, err error) {
+	tblDef, err := metricschema.GetMetricTableDef(e.table.Name.L)
```

Comment:
Why get the table definition in `getRows`? I think we should get the definition in the executor build stage.

Reply:
Ye, but the executor building stage can't return error, so I move this in here.

Reply:
How about in `Open`?

Reply:
done.